### PR TITLE
feat(react): create client tailored to React

### PR DESCRIPTION
**Pull request by balazsorban44**
**What**:

These changes ensure that we work more tightly with React that can also result in unforeseen performance boosts. In case we would decide on expanding to other libraries/frameworks, a new file per framework could be added.

**Why**:

Some performance issues (https://github.com/nextauthjs/next-auth/issues/844) could only be fixed by moving more of the client code into the `Provider`.

**How**:

Refactoring `next-auth/client`

Related: #1461, #1084, #1462

BREAKING CHANGE:
**1.** `next-auth/client` is renamed to `next-auth/react`.

**2.** In the past, we exposed most of the functions with different names for convenience. To simplify our source code, the new React specific client code exports only the following functions, listed with the necessary changes:

- `setOptions`: Not exposed anymore, use `SessionProvider` props
- `options`: Not exposed anymore, use `SessionProvider` props
- `session`: Rename to `getSession`
- `providers`: Rename to `getProviders`
- `csrfToken`: Rename to `getCsrfToken`
- `signin`: Rename to `signIn`
- `signout`: Rename to `signOut`
- `Provider`: Rename to `SessionProvider`

**3.** `Provider` changes.
- `Provider` is renamed to `SessionProvider`
- The `options` prop is now flattened as the props of `SessionProvider`.
- `clientMaxAge` has been renamed to `staleTime`.
- `keepAlive` has been renamed to `refetchInterval`.
An example of the changes:
```diff
- <Provider options={{clientMaxAge: 0, keepAlive: 0}}>{children}</Provider>
+ <SessionProvider staleTime={0} refetchInterval={0}>{children}</SessionProvider> 
```

**4.** It is now **required** to wrap the part of your application that uses `useSession` into a `SessionProvider`.

Usually, the best place for this is in your `pages/_app.jsx` file:

```jsx
import { SessionProvider } from "next-auth/react"

export default function App({
  Component,
  pageProps: { session, ...pageProps }
}) {
  return (
    // `session` comes from `getServerSideProps` or `getInitialProps`.
    // Avoids flickering/session loading on first load.
    <SessionProvider session={session}>
      <Component {...pageProps} />
    </SessionProvider>
  )
}
```

### CODE REVIEW

File: www/docs/tutorials/securing-pages-and-api-routes.md [hunk 1]
```diff
@@ -69,7 +69,7 @@ export async function getServerSideProps(context) {
 This example assumes you have configured `_app.js` to pass the `session` prop through so that it's immediately available on page load to `useSession`.
 
 ```js title="pages/_app.js"
-import { Provider } from 'next-auth/client'
+import { Provider } from "next-auth/react"
```
Comment (hboylan) [inline www/docs/tutorials/securing-pages-and-api-routes.md:72]: Looks like you've renamed `Provider` to `SessionProvider` so all documentation references need to be updated as well.

**Review by hboylan — COMMENTED** [review]
Looks good so far!

File: www/docs/tutorials/securing-pages-and-api-routes.md (hunk 1 above)
Comment (balazsorban44) [reply www/docs/tutorials/securing-pages-and-api-routes.md:72]: Exactly. This is still a WIP, but thanks for the heads up. :)

File: src/client/react.js [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+      window.addEventListener("storage", handler)
+      return () => window.removeEventListener("storage", handler)
+    },
+    /** Notify other tabs/windows. */
+    post(message) {
+      localStorage.setItem(
+        name,
+        JSON.stringify({ ...message, timestamp: _now() })
+      )
+    },
```
Comment (hboylan) [inline src/client/react.js:349]:
```suggestion
    post(message) {
      if (typeof localStorage === "undefined") return
      localStorage.setItem(
        name,
        JSON.stringify({ ...message, timestamp: _now() })
      )
    },
```

**Review by hboylan — COMMENTED** [review]
I used `npm link` for my existing project to confirm this solves the issue of the unnecessary client-side session call. 🎉 

I had to install `nodemailer`, which seems like it only be an `peerOptionalDependency` though.

File: src/client/react.js (hunk 2 above)
Comment (balazsorban44) [reply src/client/react.js:349]: My experience was that this check wasn't needed. Did you run into problems without it?

**Comment by balazsorban44** [comment]
@hboylan yes, from version 4, nodemailer will be optional, as it's only needed if someone uses the email provider. anyone else won't have to install it. smaller bundle size, faster install. 😊

**Review by 0ubbe — APPROVED** [review]
Impressive work 💯 , did a pair programming review with @balazsorban44 and in my opinion, we're good to go as a base for **v4** 👍🏽

File: src/client/react.js (hunk 2 above)
Comment (hboylan) [reply src/client/react.js:349]: I believe it's needed, but I may be doing something wrong. My `getSession` call on the server yields the error "ReferenceError: localStorage is not defined".
Screen Shot 2021-06-11 at 2 41 22 PM

Comment (balazsorban44) [reply src/client/react.js:349]: I see :grimacing: I just merged this PR, but I'll create another one with a fix.

Comment (hboylan) [reply src/client/react.js:349]: Ah, sorry for my delayed response! I'll give v4 a try now 😄

**Pull request merged**